### PR TITLE
ci(cppcheck-all, cppcheck-differential): add cppcheck

### PR DESCRIPTION
**Pull request by kminoda**
## Description
Add cppcheck for GitHub Actions
- cppcheck-differential: A workflow that is applied only to the pull request, mainly to check if the modified files are cppcheck-error-free. This only checks limited items in cppcheck to make it pass in the current codebase, aiming at setting this CI mandatory ASAP.
- cppcheck-all: A workflow that is intended to run on all files under Autoware Universe, mainly to monitor the remaining cppcheck errors. This may also be run manually as well.

Future improvements would include
- Address the errors pointed out in cppcheck-all to achieve cppcheck-error-free Autoware Universe
- Apply cppcheck not only to Autoware Universe but also to other implementations (e.g. autoware_common)
- Summarize the results of cppcheck-all to motivate each engineers to address the warnings

## Related links

- [TIER IV INTERNAL JIRA LINK](https://tier4.atlassian.net/browse/RT1-6244)

## Tests performed

See the results of GitHub Actions

## Notes for reviewers

We would eventually like to apply most of the important cppcheck items, but for now I created a huge `.cppcheck_suppression` file to suppress all the errors. 

## Interface changes

None

### ROS Topic Changes

None

### ROS Parameter Changes

None

## Effects on system behavior

None

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [ ] I've confirmed the [contribution guidelines].
- [ ] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by kminoda** [comment]
@veqcc @TakaHoribe Hi, would you review this PR?

**Comment by veqcc** [comment]
@kminoda 
Thank you for your quick actions!!

My first question is, are all the features listed in `.cppcheck_suppressions` expected to pass immediately? Or, will be mandatory in the near future, like in several months?

If you are willing to make it mandatory as soon as possible, I think we should choose the features more carefully.

**Comment by veqcc** [comment]
@kminoda 

> .cppcheck_suppressions is the list of items to "ignore" rather than to "include". So no, those features are not expected to pass immediately.

I was misunderstanding in an opposite way, thanks!

> (BTW the only risk I can see is the case when cppcheck raises false-positive errors.

Static analysis tools are always incomplete.
All tools are recommended to use several tools at the same time.

I will approve it later.
My final concern is, is there any future plan which features listed in the suppression file are to be mandatory some time or suppressed forever?

**Review by veqcc — APPROVED** [review]

File: .github/workflows/cppcheck-all.yaml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+      - name: Install dependencies
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y build-essential cmake git libpcre3-dev
+
+      # cppcheck from apt does not yet support --check-level args, and thus install from source
+      - name: Install Cppcheck from source
+        run: |
+          mkdir /tmp/cppcheck
+          git clone https://github.com/danmar/cppcheck.git /tmp/cppcheck
```
Comment (TakaHoribe) [inline .github/workflows/cppcheck-all.yaml:26]: Do we have to build from the source? Can't we use the apt-released version?

**Review by TakaHoribe — APPROVED** [review]
> The idea behind this is to suppress all (and only) the errors in the latest Autoware Universe so that we can make the CI mandatory as soon as this PR is merged. 

This sounds reasonable to me. I approve this, but please check one question below.

File: .github/workflows/cppcheck-all.yaml (hunk 1 above)
Comment (kminoda) [reply .github/workflows/cppcheck-all.yaml:26]: As written in the comment above as follow:
- cppcheck from apt does not yet support --check-level args, and thus install from source

cppcheck version 2.7, which is the version installed from apt, did not support some argument that I wanted to use. Maybe it is possible to work around with v2.7, but as for now I chose to build from source.

Comment (kminoda) [reply .github/workflows/cppcheck-all.yaml:26]: This `--check-level` argument: https://sourceforge.net/p/cppcheck/news/2023/06/cppcheck-211/

Comment (TakaHoribe) [reply .github/workflows/cppcheck-all.yaml:26]: I got it. Thank you!

Comment (HansRobo) [reply .github/workflows/cppcheck-all.yaml:26]: `cppcheck` is now binary installed by #7611 !

**Pull request merged**